### PR TITLE
Fix end_to_end tests for GPDB 7

### PR DESCRIPTION
**Pull request by jimmyyih**
This PR should fix a lot of issues in the code regarding GPDB 7 partition tables which end up fixing a lot of the end_to_end tests.  There are some other code and test fixes as well but the main thing is partitioning.

With this PR, the end_to_end tests pass against all versions of GPDB (7devel, 6X, 5X, 4.3.X).

Suggestion is to view each commit individually for review.

### CODE REVIEW

File: backup/statistics.go [hunk 1]
```diff
@@ -84,8 +88,91 @@ func GenerateAttributeStatisticsQueries(table Table, attStat AttributeStatistic)
...
+	%s::real[],
+	%s::real[],
+	%s::real[],
+	%s::real[],
+	%s,
+	%s,
+	%s,
+	%s,
+	%s`, attStat.Kind1,
+			attStat.Kind2,
```
Comment (roicos) [inline backup/statistics.go:146]: Formatting (tabs) here?

File: end_to_end/plugin_test.go [hunk 2]
```diff
@@ -80,21 +80,35 @@ var _ = Describe("End to End plugin tests", func() {
...
+			if backupConn.Version.AtLeast("7") {
+				plpythonDropStatement = "DROP PROCEDURAL LANGUAGE IF EXISTS plpython3u;"
+			}
+			testhelper.AssertQueryRuns(backupConn, plpythonDropStatement)
+			defer testhelper.AssertQueryRuns(backupConn, plpythonDropStatement)
+			defer testhelper.AssertQueryRuns(restoreConn, plpythonDropStatement)
+
 			testutils.ExecuteSQLFile(backupConn, "resources/gpdb4_objects.sql")
+			if backupConn.Version.Before("7") {
+				testutils.ExecuteSQLFile(backupConn, "resources/gpdb4_compatible_objects_before_gpdb7.sql")
```
Comment (roicos) [inline end_to_end/plugin_test.go:107]: A little bit confused is it for GPDB4 or GPDB7?

File: backup/predata_relations.go [hunk 3]
```diff
@@ -38,25 +38,59 @@ func SplitTablesByPartitionType(tables []Table, includeList []string) ([]Table,
...
+				dataTables = append(dataTables, table)
+			}
+
+			// In GPDB 7+, we need to filter out leaf and intermediate subroot partitions
+			// from being added to the metadata table list if their root partition parent
+			// is in the exclude list. This is to prevent ATTACH PARTITION statements
+			// against nonexistant root partitions from being printed to the metadata file.
+			if connectionPool.Version.AtLeast("7") &&
+				table.AttachPartitionInfo != (AttachPartitionInfo{}) &&
+				!excludeList.MatchesFilter(table.AttachPartitionInfo.Parent) {
```
Comment (roicos) [inline backup/predata_relations.go:85]: Isn't `!excludeList.MatchesFilter(table.AttachePartitionInfo.Parent)`  for skipping the table whose parent is NOT in the excludeList?

File: end_to_end/plugin_test.go (hunk 2 above)
Comment (jimmyyih) [reply end_to_end/plugin_test.go:107]: `gpdb4_compatible_objects_before_gpdb7.sql` is for GPDB 4 compatible objects that can be executed on GPDB 4, 5, and 6.  `gpdb4_compatible_objects_after_gpdb7.sql` is for GPDB 4 compatible objects that can be executed on GPDB 7+.  The one for GPDB 7+ has slight modifications to make it work for GPDB 7+.

File: backup/statistics.go (hunk 1 above)
Comment (jimmyyih) [reply backup/statistics.go:146]: It's consistent with the formatting for `generateAttributeSlotsQuery4()` and `generateAttributeSlotsQuery6()`.  Can change it with a refactoring commit for the entire file in a later refactoring PR.

File: backup/predata_relations.go (hunk 3 above)
Comment (jimmyyih) [reply backup/predata_relations.go:85]: Yeah, the set implementation in `utils/set.go` can be confusing.  Here, `excludeList.MatchesFilter(table.AttachePartitionInfo.Parent)` will return false if the parent is in the map which I then negate to true to enter the conditional and `continue`.

File: backup/backup.go [hunk 4]
```diff
@@ -471,6 +471,9 @@ func cancelBlockedQueries(timestamp string) {
 		select {
 		case <-tickerCheckCanceled.C:
 			blockedQueryCount := fmt.Sprintf("SELECT count(*) from pg_stat_activity WHERE application_name='gpbackup_%s' AND waiting='t' AND  waiting_reason='lock';", timestamp)
+			if conn.Version.AtLeast("7") {
+				blockedQueryCount = fmt.Sprintf("SELECT count(*) from pg_stat_activity WHERE application_name='gpbackup_%s' AND wait_event_type='Lock';", timestamp)
```
Comment (kyeap-vmware) [inline backup/backup.go:475]: I coulda sworn that I fixed this?!?! I thought I remember us talking about it. Also the commit message has a typo `:%s/wait_even_type/wait_event_type`

Comment (jimmyyih) [reply backup/backup.go:475]: You mean https://github.com/greenplum-db/gpbackup/pull/427/commits/20a9c895773c5c3e3f93f8babf597d08624e5978 from https://github.com/greenplum-db/gpbackup/pull/427?  Yeah, you forgot to look lower down in the code lol.

Ack on the typo.  Will fix.

File: restore/restore.go [hunk 5]
```diff
@@ -334,6 +334,18 @@ func editStatementsRedirectSchema(statements []toc.StatementWithType, redirectSc
 		newSchema := fmt.Sprintf("%s.", redirectSchema)
 		statements[i].Schema = redirectSchema
 		statements[i].Statement = strings.Replace(statement.Statement, oldSchema, newSchema, 1)
```
Comment (shivzone) [inline restore/restore.go:336]: Just for reference can you also mention the pre gpdb 7 syntax as a comment prior to the gpdb7 block

File: backup/statistics.go [hunk 6]
```diff
@@ -84,8 +88,91 @@ func GenerateAttributeStatisticsQueries(table Table, attStat AttributeStatistic)
...
+			AnyValues(attStat.Values3, attStat.Type),
+			AnyValues(attStat.Values4, attStat.Type),
+			AnyValues(attStat.Values5, attStat.Type))
+	}
+	return attributeQuery
+}
+
 // GPDB6 introduced an additional statistic slot that we account for in this function
-func generateAttributeSlotsQueryMaster(attStat AttributeStatistic) string {
+func generateAttributeSlotsQuery6(attStat AttributeStatistic) string {
```
Comment (shivzone) [inline backup/statistics.go:175]: Not relevant to this PR. In general i wish we stayed away from using references to `masterQuery` or use function names with `*master*`. master is a moving target and i see numerous references to masterQuery that are some times applied to 6+ and some 7+

Comment (jimmyyih) [reply backup/statistics.go:175]: Yeah, we can do this in a separate refactor PR after master_changes is done.

File: restore/restore.go (hunk 5 above)
Comment (jimmyyih) [reply restore/restore.go:336]: There is no before syntax.  For pre-GPDB7, the redirect schema feature would just replace the first instance of the schema name (e.g. CREATE TABLE fooschema.foobar => CREATE TABLE newschema.foobar).  Nothing specific until now where we required two replacements.

**Comment by jimmyyih** [comment]
Note that this PR makes end_to_end green for GPDB 7 only if you have a GPDB 7 with https://github.com/greenplum-db/gpdb/pull/11263 patch.  Otherwise, the statistics tests are expected failures.

**Review by kyeap-vmware — APPROVED** [review]

**Review by shivzone — APPROVED** [review]

**Review by roicos — APPROVED** [review]

**Pull request merged**